src/shogun/mathematics/graph/ops/shogun/details/add_avx.cpp

```diff
@@ -4,43 +4,54 @@
  * Authors: Gil Hoben
  */
 
-#include <unsupported/Eigen/CXX11/Tensor>
+#include <algorithm>
+#include <functional>
+#include <Eigen/Core>
+#include "PacketType.h"
+
+using namespace Eigen::internal;
 
 namespace shogun::graph::op {
+
+	struct Packet
+	{
+		aligned_vector m_data;
+		const size_t m_size;
+	};
+
 	template <typename T>
 	void add_kernel_implementation_avx(
-	    void* input1, void* input2, void* output, const size_t size);
+	    void* input1, void* input2, void* output);
 
 	template <typename T>
 	void add_kernel_implementation_avx(
-	    void* input1, void* input2, void* output, const size_t size)
+	    void* input1, void* input2, void* output)
 	{
-		Eigen::TensorMap<Eigen::Tensor<T, 1>> A(static_cast<T*>(input1), size);
-		Eigen::TensorMap<Eigen::Tensor<T, 1>> B(static_cast<T*>(input2), size);
-		Eigen::TensorMap<Eigen::Tensor<T, 1>> Out(static_cast<T*>(output), size);
-
-		Out.device(Eigen::DefaultDevice{}) = A + B;
+		std::transform(
+		    static_cast<const T*>(input1),
+		    static_cast<const T*>(input1) + 32/sizeof(T),
+		    static_cast<const T*>(input2), static_cast<T*>(output),
+		    std::plus<T>());
 	}
 
-	template <>
-	void add_kernel_implementation_avx<bool>(
-	    void* input1, void* input2, void* output, const size_t size)
+	template<>
+	void add_kernel_implementation_avx<float>(
+		void* input1, void* input2, void* output)
 	{
-		std::transform(
-		    static_cast<const bool*>(input1),
-		    static_cast<const bool*>(input1) + size,
-		    static_cast<const bool*>(input2), static_cast<bool*>(output),
-		    std::plus<bool>());
+		static_cast<Packet*>(output)->m_data = padd(
+			std::get<Packet8f>(static_cast<const Packet*>(input1)->m_data), 
+			std::get<Packet8f>(static_cast<const Packet*>(input2)->m_data));
```

Comment:
@karlnapf @vigsterkr the packet data lives on the stack because the variant is basically a union so it can store data on stack instead of heap. This means that we can "type erase" __m128 and so on.. :)